src/lib.rs

```diff
@@ -690,22 +693,26 @@ unsafe fn gen_picture(c: &mut Rav1dContext) -> Rav1dResult {
     if output_picture_ready(c, false) {
         return Ok(());
     }
-    while c.in_0.sz > 0 {
-        let r#in = mem::take(&mut c.in_0); // Take so we don't have 2 `&mut`s.
-        let len = rav1d_parse_obus(c, &r#in, false);
-        c.in_0 = r#in; // Restore into `c` right after.
-        match len {
-            Err(_) => rav1d_data_unref_internal(&mut c.in_0),
-            Ok(len) => {
-                assert!(len <= c.in_0.sz);
-                c.in_0.sz -= len;
-                c.in_0.data = c.in_0.data.add(len);
-                if c.in_0.sz == 0 {
-                    rav1d_data_unref_internal(&mut c.in_0);
-                }
-            }
+    // Take so we don't have 2 `&mut`s.
+    let Rav1dData {
+        data: r#in,
+        m: props,
+    } = mem::take(&mut c.in_0);
+    let Some(mut r#in) = r#in else { return Ok(()) };
+    while !r#in.is_empty() {
+        let len = rav1d_parse_obus(c, &r#in, &props, false);
+        if let Ok(len) = len {
+            r#in.slice_in_place(len..);
         }
+        // Note that [`output_picture_ready`] doesn't read [`Rav1dContext::in_0`].
```

Comment:
general nit (not something to necessarily fix here, just a note for the future): I wouldn't bother using the `[markdown link syntax]` in non-doc comments. Keep using \`backticks\` around item names, but the [square brackets] just add noise when they're not actually being used to create a link in the generated docs.